### PR TITLE
Implement [ink::test] proc macro

### LINKED ISSUE #486: Add `#[ink::test]` procedural macro to reduce code duplication
## Problem

We currently have a lot of code duplication for ink! core tests:

```rust
#[test]
fn spread_layout_works() -> env::Result<()> {
    env::test::run_test::<env::DefaultEnvTypes, _>(|_| {
       // test code returning a Result<(), env::Error>
    })
}
```

In the past we sometimes had the case of having to adapt all these tests, because of some change to this setup.

## Proposal

Add a procedural macro à la `#[ink::test]` to mitigate this:

```rust
use ink_lang as ink;

#[ink::test]
fn spread_layout_works() -> env::Result<()> {
   // test code that returns Result<(), env::Error>
}
```

or 

```rust
use ink_lang as ink;

#[ink::test]
fn spread_layout_works() {
   // testing code that returns nothing
}
```

### PR DESCRIPTION
**Pull request by cmichi**
Closes #486.

### CODE REVIEW

File: lang/macro/src/codegen/ink_test.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
```
Comment (Robbepop) [inline lang/macro/src/codegen/ink_test.rs:1]: Incorrect date.

File: lang/macro/src/codegen/ink_test.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        storage::Storage,
+        GenerateCode,
+        GenerateCodeUsing,
+    },
+    ir::InkTest,
+};
+
+/// Generates code for the `[ink::test]` macro.
+#[derive(From)]
+pub struct InkTestModule<'a> {
```
Comment (Robbepop) [inline lang/macro/src/codegen/ink_test.rs:38]: The name should be just `InkTest` since I do not even know what "module" you are referring to.

File: lang/macro/src/codegen/ink_test.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    /// Generates the code for `#[ink:test]`.
+    fn generate_code(&self) -> TokenStream2 {
+        let item_fn = &self.ink_test.item_fn;
+        let attrs = &item_fn.attrs;
+        let sig = &item_fn.sig;
+        let fn_name = &sig.ident;
+        let fn_return_type = &sig.output;
+        let fn_block = &item_fn.block;
+        let _vis = &item_fn.vis;
+        let _fn_args = &sig.inputs;
```
Comment (Robbepop) [inline lang/macro/src/codegen/ink_test.rs:53]: Never silently ignore broken inputs and do nothing with them. Either tell the user that visibility modifiers and function inputs are forbidden in test functions or simply re-generate them so that the rust compiler can take over this burden from you.
tl;dr: Don't try to generate errors or warnings and simply re-generate the degenerated function and let the Rust compiler do its job.

File: lang/macro/src/codegen/ink_test.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        match fn_return_type {
```
Comment (Robbepop) [inline lang/macro/src/codegen/ink_test.rs:54]: Technically you could also have the case of a `-> ()` return signature that is obviously equal to having no specified return value. This is a bit tricky to match against though.

File: lang/macro/src/codegen/ink_test.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            syn::ReturnType::Default => {
+                quote! {
+                    #( #attrs )*
+                    #[test]
+                    fn #fn_name() {
+                        env::test::run_test::<env::DefaultEnvTypes, _>(|_| {
+                            #fn_block
```
Comment (Robbepop) [inline lang/macro/src/codegen/ink_test.rs:61]:
```suggestion
                            { #fn_block }
```

File: lang/macro/src/codegen/ink_test.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                    }
+                }
+            }
+            syn::ReturnType::Type(_rarrow, _type) => {
+                quote! {
+                    #( #attrs )*
+                    #[test]
+                    fn #fn_name() -> env::Result<()> {
+                        env::test::run_test::<env::DefaultEnvTypes, _>(|_| {
+                            #fn_block
```
Comment (Robbepop) [inline lang/macro/src/codegen/ink_test.rs:74]:
```suggestion
                            { #fn_block }
```

File: lang/macro/src/codegen/ink_test.rs (hunk 5 above)
Comment (Robbepop) [reply lang/macro/src/codegen/ink_test.rs:61]: Also: Have you tested that this actually produces the expected code? I'd be very curious since a function block is normally the actual function body and what you do here is that you embed this function body into yet another newly generated function body. What I would do instead is to carve out the statements of the original function body (get an iterator to the statements or so) and then generate them using `#( #statements )*` syntax.

File: lang/macro/src/codegen/ink_test.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            syn::ReturnType::Default => {
+                quote! {
+                    #( #attrs )*
+                    #[test]
+                    fn #fn_name() {
+                        env::test::run_test::<env::DefaultEnvTypes, _>(|_| {
+                            #fn_block
+                            Ok(())
+                        })
+                        .expect(&format!("{}: the off-chain testing environment returned an error", stringify!(#fn_name)));
```
Comment (Robbepop) [inline lang/macro/src/codegen/ink_test.rs:64]: You do not have to make use of the `format` macro here. Since we are in a proc. macro we can simply pre-generate the format string ourselves using `format!` in the proc. macro implementation and simply insert the string contents in the `expect` method as static str.

File: lang/macro/src/codegen/ink_test.rs (hunk 6 above)
Comment (Robbepop) [reply lang/macro/src/codegen/ink_test.rs:74]: Also I think in this case where the user wants to return a `Result` we do not need to append the `Ok(())` in order to make their lifes easier because it won't.

File: lang/macro/src/codegen/ink_test.rs (hunk 5 above)
Comment (Robbepop) [reply lang/macro/src/codegen/ink_test.rs:61]: Also: We want to simulate for users that they are actually writing code in a function that does not return `Result` but instead nothing. However with the currently generated code they could still make use of `?`-operator et. al. So you need to find a way to surpress this and make it look'n'feel to users as if they were really just writing code in a function that returns nothing. E.g. wrapping the input statements in an embedded internal function that actually returns nothing and simply call that one in the actual function body followed by `Ok(())`.

Comment (Robbepop) [reply lang/macro/src/codegen/ink_test.rs:61]: Or even simpler, generate something like:
```rust
{
    let _: () = {
        #( #statements )*
    };
    Ok(())
}
```
as function body

File: lang/macro/src/ir/into_hir.rs [hunk 8]
```diff
@@ -56,6 +56,14 @@ impl Parse for ir::Marker {
     }
 }
 
+impl TryFrom<(ir::Params, syn::ItemFn)> for ir::InkTest {
+    type Error = syn::Error;
+
+    fn try_from((_params, item_fn): (ir::Params, syn::ItemFn)) -> Result<Self> {
+        Ok(Self { item_fn })
+    }
+}
```
Comment (Robbepop) [inline lang/macro/src/ir/into_hir.rs:65]: Why not simply make it
```rust
impl TryFrom<syn::ItemFn> for ir::InkTest { ... }
```
?

**Review by Robbepop — CHANGES_REQUESTED** [review]

File: lang/macro/src/codegen/ink_test.rs (hunk 4 above)
Comment (cmichi) [reply lang/macro/src/codegen/ink_test.rs:54]: I had thought of the same thing, but the [documentation](https://docs.rs/syn/1.0.41/syn/enum.ReturnType.html#variant.Default) luckily states:
```
Default
Return type is not specified.
Functions default to () and closures default to type inference.
```

File: lang/macro/src/codegen/ink_test.rs (hunk 6 above)
Comment (cmichi) [reply lang/macro/src/codegen/ink_test.rs:74]: Can you elaborate how it won't make their life easier? If we omit the `Ok(())` here they would have to add it manually in every test.

Comment (Robbepop) [reply lang/macro/src/codegen/ink_test.rs:74]: it is more transparent as to what they would write if their test function was returning a `Result`. after all we should only generate barely enough magic as to make this still appear transparent enough for users to derive a sense of it. I would definitely be baffled by having an implicit return of `Ok((())` that I do not see myself.

File: lang/macro/src/codegen/ink_test.rs (hunk 4 above)
Comment (Robbepop) [reply lang/macro/src/codegen/ink_test.rs:54]: Yeah but what if a user wrote: `fn foo() -> ()`. That is valid and `syn` will then not make the return type `Default` but instead will report an empty tuple as return type.

File: lang/macro/src/codegen/ink_test.rs (hunk 5 above)
Comment (cmichi) [reply lang/macro/src/codegen/ink_test.rs:61]: > However with the currently generated code they could still make use of ?-operator et. al. 

Ah, good catch!

File: lang/macro/src/codegen/ink_test.rs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    env_types::EnvTypes,
+    events::{
+        EventHelpers,
+        EventStructs,
+    },
+    metadata::GenerateMetadata,
+    storage::Storage,
+    GenerateCode,
+    GenerateCodeUsing,
+};
```
Comment (Robbepop) [inline lang/macro/src/codegen/ink_test.rs:31]: Please make sure that you do use all the imports.

File: lang/macro/src/ink_test.rs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    lint::idents_respect_pred(
+        input.clone(),
+        move |ident| !ident.to_string().starts_with("__ink"),
+        move |ident| {
+            format_err_spanned!(
+                ident,
+                "identifiers starting with `__ink` are forbidden in ink!"
+            )
+        },
+    )?;
```
Comment (Robbepop) [inline lang/macro/src/ink_test.rs:42]: FYI: This lint algorithm is actually deprecated but doesn't matter too much for the sake of this PR. So just leave it.

File: lang/macro/src/ir/data.rs [hunk 11]
```diff
@@ -24,6 +24,7 @@ use quote::ToTokens;
 use syn::{
     punctuated::Punctuated,
     spanned::Spanned as _,
+    ItemFn,
```
Comment (Robbepop) [inline lang/macro/src/ir/data.rs:27]: In the other codebases we normally do not directly import `syn` items and instead use their short `syn` prefix to make it clear in the code that these are `syn` types. This is especially useful when ink! shares a lot of common names. Not particularly useful for this small macro though. So `ir::` is always from ink! and `syn::` is always from ... well `syn`.

**Review by Robbepop — APPROVED** [review]
All in all looks good to me!
